cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -1072,6 +1133,74 @@ private ProgramId validateProgramForRunRecord(String namespaceName, String appNa
   }
 
   /**
+   * Helper method to update log levels for Worker, Flow, or Service.
+   */
+  private void updateLogLevels(ProgramId programId, Map<String, String> logLevels, @Nullable String component)
+    throws Exception {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId);
+    if (runtimeInfo != null) {
+      TwillController twillController = getTwillController(runtimeInfo);
+      if (component == null) {
+        twillController.updateLogLevels(transformLogLevelsMap(logLevels)).get();
+      } else {
+        twillController.updateLogLevels(component, transformLogLevelsMap(logLevels)).get();
+      }
+    }
+  }
+
+  /**
+   * Helper method to reset log levels for Worker, Flow or Service.
+   */
+  private void resetLogLevels(ProgramId programId, @Nullable String[] loggerNames, @Nullable String component)
+    throws Exception {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId);
+    if (runtimeInfo != null) {
+      TwillController twillController = getTwillController(runtimeInfo);
+      if (component == null) {
+        if (loggerNames == null || loggerNames.length == 0) {
+          twillController.resetLogLevels().get();
+        } else {
+          twillController.resetLogLevels(loggerNames).get();
+        }
+      } else {
+        if (loggerNames == null || loggerNames.length == 0) {
+          twillController.resetRunnableLogLevels(component).get();
+        } else {
+          twillController.resetRunnableLogLevels(component, loggerNames).get();
+        }
+      }
+    }
+  }
+
+  /**
+   * Helper method to get the {@link TwillController} for the program.
+   */
+  private TwillController getTwillController(ProgramRuntimeService.RuntimeInfo runtimeInfo) throws Exception {
+    ProgramController programController = runtimeInfo.getController();
+    if (!(programController instanceof AbstractTwillProgramController)) {
+      throw new BadRequestException("Update log levels at runtime is only supported in distribute mode");
+    }
+    return ((AbstractTwillProgramController) programController).getTwillController();
```

Comment:
rather than checking the instance here and creating a backdoor to get the twill controller, I think it would be cleaner to add methods to the ProgramController to set and reset the log level. Only the distributed implementation actually does something with it, while the other implementations throw an UnsupportedOperationException